src/codemirror/utils/generate-autocompletion.ts

```diff
@@ -0,0 +1,90 @@
+import {
+  DefaultValueDescriptor,
+  Documentation,
+  PropDescriptor,
+} from "react-docgen/dist/Documentation";
+const excludedProps = ["children", "data-testid", "dataTestId", "id"];
+
+type ReactDocGenOutput = Record<string, Documentation[]>;
+type PropEntry = [string, PropDescriptor];
+type TsPropDescriptor = PropDescriptor["tsType"];
+type JsPropDescriptor = PropDescriptor["type"];
+
+export function generateAutocompletion(input: ReactDocGenOutput) {
+  return Object.values(input)
+    .flat()
+    .reduce((acc, { displayName, props = {} }) => {
+      return {
+        ...acc,
+        [displayName]: Object.entries(props).filter(filterProps).map(mapProps),
+      };
+    }, {});
+}
+
+function filterProps([propName]: PropEntry): boolean {
+  return !excludedProps.includes(propName);
+}
+
+function mapProps([propName, propValue]: PropEntry) {
+  const {
+    tsType,
+    type: jsType,
+    required = false,
+    description = "",
+    defaultValue,
+  } = propValue;
+
+  return {
+    [propName]: {
+      type: parseType(tsType ?? jsType),
+      required,
+      defaultValue: parseDefaultValue(defaultValue),
+      description,
+    },
+  };
+}
+
+function parseType(propType: JsPropDescriptor | TsPropDescriptor): unknown {
+  if (!propType) {
+    return "unknown";
+  }
+
+  switch (propType.name) {
```

Comment:
Isn't it a bit low level? Is there a library maybe for this purpose?

Reply:
basically it is what `react-docgen` outputs, so unfortunately no, did not find a util that handles `react-docgen`'s result 😞

Reply:
If there is one, I'll be mad at the amount of time I've spent on it 😅